### PR TITLE
Swiftmailer - Fixes #401

### LINKED ISSUE #401: Register SwiftMailer as a service
Register the SwiftMailer as a service so it can be used throughout Fork.

Todo:

- [x] Register swiftMailer as as service
- [x] Extend Swift_Message with some fork specific functionality and dependency injection (f.e. Fork_Message)
- [x] Fork_Message: template parsen, talen automatisch assignen
- [x] Logging with monolog
- [x] Deprecate commonMailer
- [x] Refactor usages of CommonMailer

### PR DESCRIPTION
**Pull request by WouterSioen**
SpoonEmail is not used anymore. Swiftmailer is now used to send the emails.

The addEmail method is deprecated in favor of a send method that accepts a Swift_Message object (or the extended Fork version: Common\Mailer\Message.

Usage:

    // create a message object and set all the needed properties
    $message = \Common\Mailer\Message::newInstance($subject)
        ->setFrom(array($fromEmail => $fromName))
        ->setTo(array($toEmail => $toName))
        ->setReplyTo(array($replyToEmail => $replyToName))
        ->parseHtml($template, $variables, $addUTM)
        ->setPlainText($plainText)
        ->addAttachments($attachments)
    ;

    // send it trough the mailer service
    $this->get('mailer')->send($message);

Note that the old addEmail method isn't removed. The $queue en $isRawHtml properties aren't respected anymore though. Support for these two features is dropped because it wasn't used anymore.

### CODE REVIEW

**Comment by WouterSioen** [comment]
Todo: log before sending emails.

File: src/Frontend/Modules/Profiles/Actions/ResendActivation.php [hunk 1]
```diff
@@ -121,14 +121,21 @@ private function validateForm()
...
+                $from = FrontendModel::getModuleSetting('Core', 'mailer_from');
+                $replyTo = FrontendModel::getModuleSetting('Core', 'mailer_reply_to');
+                $message = \Common\Mailer\Message::newInstance(
+                        FL::getMessage('RegisterSubject')
+                    )
+                    ->setFrom(array($from['email'] => $from['name']))
+                    ->setTo(array($profile->getEmail() => ''))
+                    ->setReplyTo(array($replyTo['email'] => $replyTo['name']))
+                    ->parseHtml(
+                        FRONTEND_MODULES_PATH . '/Profiles/Layout/Templates/Mails/register.tpl',
```
Comment (Diewy) [inline src/Frontend/Modules/Profiles/Actions/ResendActivation.php]: Wrong filename

File: src/Frontend/Modules/Profiles/Actions/Register.php [hunk 2]
```diff
@@ -169,14 +169,21 @@ private function validateForm()
...
+                    $from = FrontendModel::getModuleSetting('Core', 'mailer_from');
+                    $replyTo = FrontendModel::getModuleSetting('Core', 'mailer_reply_to');
+                    $message = \Common\Mailer\Message::newInstance(
+                            FL::getMessage('RegisterSubject')
+                        )
+                        ->setFrom(array($from['email'] => $from['name']))
+                        ->setTo(array($txtEmail->getValue() => ''))
+                        ->setReplyTo(array($replyTo['email'] => $replyTo['name']))
+                        ->parseHtml(
+                            FRONTEND_MODULES_PATH . '/Profiles/Layout/Templates/Mails/register.tpl',
```
Comment (Diewy) [inline src/Frontend/Modules/Profiles/Actions/Register.php]: Wrong filename

File: src/Frontend/Modules/Profiles/Actions/ForgotPassword.php [hunk 3]
```diff
@@ -122,14 +122,21 @@ private function validateForm()
...
+                $from = FrontendModel::getModuleSetting('Core', 'mailer_from');
+                $replyTo = FrontendModel::getModuleSetting('Core', 'mailer_reply_to');
+                $message = \Common\Mailer\Message::newInstance(
+                        FL::getMessage('ForgotPasswordSubject')
+                    )
+                    ->setFrom(array($from['email'] => $from['name']))
+                    ->setTo(array($txtEmail->getValue() => ''))
+                    ->setReplyTo(array($replyTo['email'] => $replyTo['name']))
+                    ->parseHtml(
+                        FRONTEND_MODULES_PATH . '/Profiles/Layout/Templates/Mails/forgot_password.tpl',
```
Comment (Diewy) [inline src/Frontend/Modules/Profiles/Actions/ForgotPassword.php]: Wrong filename

File: src/Frontend/Modules/Faq/Widgets/AskOwnQuestion.php [hunk 4]
```diff
@@ -117,15 +117,22 @@ private function validateForm()
...
+                $from = FrontendModel::getModuleSetting('Core', 'mailer_from');
+                $replyTo = FrontendModel::getModuleSetting('Core', 'mailer_reply_to');
+                $message = \Common\Mailer\Message::newInstance(
+                        sprintf(FL::getMessage('FaqOwnQuestionSubject'), $variables['name'])
+                    )
+                    ->setFrom(array($from['email'] => $from['name']))
+                    ->setTo(array($variables['email'] => $variables['name']))
+                    ->setReplyTo(array($replyTo['email'] => $replyTo['name']))
+                    ->parseHtml(
+                        FRONTEND_MODULES_PATH . '/Faq/Layout/Templates/Mails/own_question.tpl',
```
Comment (Diewy) [inline src/Frontend/Modules/Faq/Widgets/AskOwnQuestion.php]: Wrong filename

File: src/Frontend/Modules/Faq/Actions/Detail.php [hunk 5]
```diff
@@ -267,13 +267,24 @@ private function validateForm()
...
+                        $from = FrontendModel::getModuleSetting('Core', 'mailer_from');
+                        $replyTo = FrontendModel::getModuleSetting('Core', 'mailer_reply_to');
+                        $message = \Common\Mailer\Message::newInstance(
+                                sprintf(FL::getMessage('FaqFeedbackSubject'), $this->record['question'])
+                            )
+                            ->setFrom(array($from['email'] => $from['name']))
+                            ->setTo(array($to['email'] => $to['name']))
+                            ->setReplyTo(array($replyTo['email'] => $replyTo['name']))
+                            ->parseHtml(
+                                FRONTEND_MODULES_PATH . '/Faq/Layout/Templates/Mails/feedback.tpl',
```
Comment (Diewy) [inline src/Frontend/Modules/Faq/Actions/Detail.php]: Wrong filename

**Comment by Diewy** [comment]
Once again, nice work! :+1:

I highlighted some wrong filenames, should be a quick fix :-)

**Comment by WouterSioen** [comment]
Fixed! I also removed the sendQueuedEmails cronjob.

**Comment by ghazispeakserve** [comment]
@WouterSioen Do you have any idea will this be merge to master?

Did all test run successfully? If yes think I will be merging it to one of our project and give a feedback if I find something.

I also wanted to share this link http://symfony.com/doc/current/cookbook/email/dev_environment.html which can be useful for debugging emails.

**Comment by WouterSioen** [comment]
Hi Ghazi,

The dev_environment link is not valid with our setup. This uses the SwiftMailerBundle from Symfony. We couldn't use this one though, since our mail settings are configured from the backend and not from the parameters.yml file.

This will be merged in master, yes. It's possible that there could be some more changes to the PR, but we're definitely moving to SwiftMailer.

I didn't add tests yet for this PR, since your testing PR isn't merged yet, and I didn't want to create conflicts.

**Comment by ghazispeakserve** [comment]
Thank you Wouter. It would be great if we can use the SwiftMailerBundle instead of the library.

I did not forget about my pending PRs. I am completely over-booked and I will try to find time to polish my PRs this week-end.

**Comment by WouterSioen** [comment]
It is in fact possible to change the swiftmailerbundle settings dynamically using the settings from our database, but this would be messy, since the mail configuration in parameters.yml/config.yml would be different then the configuration that will be used when sending an email. This can make it really hard to debug.

**Comment by WouterSioen** [comment]
@forkcms/developers I need your opinions: what would you prefer as best solution:

* Current implementation (step between Fork implementation and SwiftMailer)
  * Pro: (kind of) backwards compatible (custom modules won't break)
  * Con: doesn't use the Symfony bundle (yet)
* Symfony implementation (with the SwiftMailerBundle)
  * Pro: exactly the same usage as Symfony
  * Pro: easy to send debug mails trough monolog with the Swift mailer handler
  * Con: not backwards compatible
  * Con: needs an event listener to set the right transport based on backend settings.

I now went for the first implementation with the thought of refactoring to the second step later (when removing the deprecated addEmail method). Fixing the debug mails would be so much easier with the SwiftMailerBundle though.

**Comment by WouterSioen** [comment]
@forkcms/developers no input on this?

**Comment by tijsverkoyen** [comment]
I know it is a bit late, but I think I would go with the second approach. As the first way isn't fully backwards compatibel. If you are going with the first approach I would remove the possibility to configure the credentials/server thru the backend, as this is something that only the God-user can do (see: https://github.com/forkcms/forkcms/blob/master/src/Backend/Modules/Settings/Actions/Email.php#L72).

Other feedback:
* add an entry in the UPGRADE_3.9.md-file
* mark the $queue and $isRawHTML as deprecated in the next point-release (aka: current master)

**Comment by WouterSioen** [comment]
@tijsverkoyen not only the $queue and $isRawHTML will break if we go for the swiftmailerbundle. The addEmail method will be fully removed.

**Comment by tijsverkoyen** [comment]
ok, mark it as deprecated in the current master, and it can be removed in the next major-release.

**Comment by WouterSioen** [comment]
Isn't deprecating a function without a replacement a bit strange?

**Comment by tijsverkoyen** [comment]
Not really, as the UPGRADE_xxx.md will explain them on how to fix it.

**Comment by WouterSioen** [comment]
@WouterSioen add UPGRADE Guide, merge and communicate.

Swiftmailerbundle will be done in 3.9

**Pull request merged**